### PR TITLE
feat: `tx sign` and new signer

### PR DESCRIPTION
**Pull request by willemneal**
The new `sign` subcommand allows for signing a passed transaction envelope.

Additionally, `signer::{Transaction, TransactionHash}` traits which simplifies the interface. `Transaction` has just a single method which returns a `DecoratedSignature`. `TransactionHash` breaks this up into two methods to get a `Signature` and a `SignatureHint`.

### CODE REVIEW

File: cmd/soroban-cli/src/signer/types.rs
Comment (elizabethengelman) [inline cmd/soroban-cli/src/signer/types.rs:173]: I noticed a potential issue with this functionality while trying to add it to the `sign_with_lab` [branch](https://github.com/AhaLabs/stellar-cli/pull/25#discussion_r1755504162). Also possible that I am missing something or calling the command incorrectly.

When I am piping my tx xdr into the cli, the program seems to not have access to the stdin reader when it gets to trying to read the user's y response. In other words, when I run this command:
```
echo AAAAAgAAAADcOHnq5sGLOngOCEMyLqqn5CvFV2HGbOSjJAIzhqBdkAAAAGQAEb7UAAAAAQAAAAAAAAAAAAAAAQAAAAAAAAAYAAAAAQAAAAAAAAAAAAAAANw4eermwYs6eA4IQzIuqqfkK8VXYcZs5KMkAjOGoF2Q04SY/PMseMCgfbEFkX5gTr9qD/d0mvD9H3acFOz/X5oAAAAALtGgcgsAibk5q8VOLY2R1G1NdGX7fLzlz7iBFBy64JIAAAAAAAAAAAAAAAA= \
| stellar tx sign --network testnet --sign-with-key testnet-alice
```

I'm getting the same error that I'm seeing over on the `sign_with_lab` PR:
```
thread 'main' panicked at cmd/soroban-cli/src/signer.rs:164:41:
called `Result::unwrap()` on an `Err` value: Custom { kind: Other, error: "Failed to initialize input reader" }
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

File: cmd/soroban-cli/src/signer/types.rs
Comment (leighmcculloch) [inline cmd/soroban-cli/src/signer/types.rs:77]: This log line feels out of place now that we've moved to using the Print object. There'd need to be someway to pass the print object down to print here, or better move this print to a higher place. Printing here is probably not the right layer.

File: cmd/soroban-cli/src/signer/types.rs
Comment (leighmcculloch) [inline cmd/soroban-cli/src/signer/types.rs:81]: I think the trait mechanics need tweaking:

The signer defines both the signature, and the hint included in the decorated signature. The hint is not the source account, it's the signer.

So we don't need source account passed down.

The signer implementing this trait will be able to specify its own hint.

We should also remove the Blob trait, because it offers little value. Any signer to produce DecoratedSignatures would need to implement TransactionHash so they could also define the hint. No blanket impl of Blob -> TransactionHash could guess the hint.

**Review by leighmcculloch — CHANGES_REQUESTED** [review]
A few issues with this command, elaborated inline:
- Signing shouldn't require specifying a source account.
- Signing shouldn't be able to alter the source account.
- Signing shouldn't require additional user confirmation, given the command's only purpose is to sign.
- Trait mechanics. See inline. I think @Ifropc might have commented on this in the past too.

File: cmd/soroban-cli/src/signer/types.rs
Comment (elizabethengelman) [reply cmd/soroban-cli/src/signer/types.rs:173]: > The nice thing about the design of the tx sign UI is there is no need to ask for y. The command already contains a very clear authorization to sign. The user is literally stating --sign-with-key and so additional confirmation is not required imo.

This makes sense to me. 👍  I think this could be the case for most of (all of?) the `sign-with-*` options.

File: cmd/soroban-cli/Cargo.toml
Comment (willemneal) [reply cmd/soroban-cli/Cargo.toml]: Removed it from this PR since as you said it doesn't make sense to approve if we are using the sign command.

File: cmd/soroban-cli/src/config/sign_with.rs
Comment (willemneal) [reply cmd/soroban-cli/src/config/sign_with.rs]: This was a mistake on my part it's been removed.

File: cmd/soroban-cli/src/signer/types.rs
Comment (willemneal) [reply cmd/soroban-cli/src/signer/types.rs]: This has also been moved to the top level signer and not in the default implementation. In fact there is no longer a default implementation rather a blanket impl.

File: cmd/soroban-cli/src/signer/types.rs
Comment (willemneal) [reply cmd/soroban-cli/src/signer/types.rs]: Removed 👍

**Review by leighmcculloch — APPROVED** [review]
I've pushed a series of changes that I think simplify the type system around the new logic being added. The new logic is largely unchanged, just moved out of the trait system.

I was going to leave a comment with the suggestion, but wanted to try it out before suggesting, and then discovered some things along the way.

The changes involved in brief:
- Removing traits `Stellar`, `TransactionHash`, `Transaction`. Part of the design of the `StellarSigner` that was already in this PR was an enum containing the different keys, and it then called each function on its key type anyway. So we actually don't need the implementations of different keys to sit behind a common trait, at least not yet with what's presented in this PR. Also the traits were implemented at two layers, which was confusing. I think we can bring back traits when they add value, but it seems like the design of StellarSigner already satisfies.
- Removing terms like "accounts" in any of the signing logic, since these aren't accounts, just keys.
- Updated the signing code used by other fns to call through to the StellarSigner. Doing this only required ~4 lines of code because of how the StellarSigner was already designed, and it means after this PR merges we still only have one signing implementation for txs.

I'm approving this PR, but now that I've contributed changes, it would be ideal if @fnando you could 👀 the PR as a whole and sign off before we merge.

File: cmd/soroban-cli/src/config/sign_with.rs
Comment (fnando) [inline cmd/soroban-cli/src/config/sign_with.rs:33]: this is quite verbose. Should we introduce a short flag like `-s`?

File: cmd/soroban-cli/src/config/sign_with.rs
Comment (fnando) [inline cmd/soroban-cli/src/config/sign_with.rs:41]: also verbose. Should we add `-l` for signing with the lab?

File: cmd/soroban-cli/src/config/sign_with.rs
Comment (willemneal) [reply cmd/soroban-cli/src/config/sign_with.rs:33]: We have prematurely added short flags in the past and who is to say that `-s` could mean sign with ledger?

File: cmd/crates/soroban-test/tests/it/integration/tx.rs
Comment (willemneal) [reply cmd/crates/soroban-test/tests/it/integration/tx.rs:65]: By default prints are suppressed so this is just if you are interested in the output if these test. At the time it was just for debugging purposes.

File: cmd/soroban-cli/src/config/sign_with.rs
Comment (willemneal) [reply cmd/soroban-cli/src/config/sign_with.rs:41]: As I said it the other comment we can always add this later if it is requested. We could also add an alias `--with-lab`. I do think it's okay to be a little verbose as it makes the scripts more readable and you don't have to go look up the short flags if you forget.

**Comment by willemneal** [comment]
@leighmcculloch Thanks for the clean up. I do think it's ready to be merged and we can iterate when the demands change.

File: cmd/soroban-cli/src/config/sign_with.rs
Comment (leighmcculloch) [reply cmd/soroban-cli/src/config/sign_with.rs:33]: We can always add short flags later if it turns out folks really need it. Lets start with self describing commands and see how they are adopted.

Alternatively we could reduce it to `--key`.

Originally when we were adding the `--sign-with-*` we were thinking about how to introduce the sign options on other commands. But now it seems we are only doing custom signing on this command so `--sign-with-*` does seem repetitive given the command is called `tx sign`.

Comment (leighmcculloch) [reply cmd/soroban-cli/src/config/sign_with.rs:33]: I'm removing the suggestion to go with `--key`. Going with `--sign-with-key` gives us the option to add that same optino to other commands in the future, so I think we should stick with it. We can add aliases and shorten it later if we decide this was the wrong call.

File: cmd/soroban-cli/src/signer.rs
Comment (Ifropc) [inline cmd/soroban-cli/src/signer.rs:248]: Do we not intend to use prompt property here for now?

**Review by Ifropc — APPROVED** [review]
Like the new PR, very compact and simple!

File: cmd/soroban-cli/src/signer.rs
Comment (leighmcculloch) [reply cmd/soroban-cli/src/signer.rs:248]: It'll be used in a future change. We could remove it now though. I'll remove itand we can bring it back.

**Pull request merged**
